app/models/admin_user.rb

```diff
@@ -23,11 +27,54 @@ class AdminUser < ActiveRecord::Base
   validates_format_of :password, with: PASSWORD_REGEX, allow_blank: true
   validates_inclusion_of :role, in: ROLES
 
+  # = Callbacks =
+  before_update if: :crypted_password_changed? do
+    self.force_password_reset = false
+    self.password_changed_at = Time.current
+  end
+
   # = Finders =
   scope :by_name, -> { order(:last_name, :first_name) }
   scope :by_role, ->(role) { where(role: role).order(:id) }
 
   # = Methods =
+  def current_password
+    defined?(@current_password) ? @current_password : nil
+  end
+
+  def current_password=(value)
+    @current_password = value
+  end
+
+  def update_with_password(attrs)
+    if attrs[:password].blank?
+      attrs.delete(:password)
+      attrs.delete(:password_confirmation) if attrs[:password_confirmation].blank?
+    end
+
+    self.attributes = attrs
+    self.valid?
+
+    if current_password.blank?
+      errors.add(:current_password, :blank)
+    elsif !valid_password?(current_password)
+      errors.add(:current_password, :invalid)
+    elsif current_password == password
+      errors.add(:password, :taken)
+    end
+
+    errors.empty? && save(validate: false)
+  end
+
+  def destroy(current_user: nil)
```

Comment:
I've always been wary about overriding methods provided by rails (and then calling super) on ActiveRecord objects as that's historically been the source of some unexpected behaviour.

Could we use a `before_destroy` hook here, or alternatively, move the destroy logic into an action/command/service object (if that's a pattern already in use in this project) to keep the model light?